File: contracts/adapter_interface/providerAdapter.ts
// adapter_interface/providerAdapter.ts
export type InvokeInput = {
  messages?: any[];
  system?: string;
  tools?: any[];
}

export type InvokeRequest = {
  provider: string;               // 'codex' | 'openai' | ...
  operation: 'chat'|'embed'|'vision'|'audio';
  model?: string;
  budget_tokens?: number;
  temperature?: number;
  input: InvokeInput;
  metadata?: Record<string, any>;
}

export type Usage = {
  input_tokens?: number;
  output_tokens?: number;
  cost?: number;
}

export type InvokeOutput = {
  messages?: any[];
  text?: string;
  tool_calls?: any[];
}

export type InvokeResponse = {
  output: InvokeOutput;
  usage: Usage;
  provider?: string;
  model?: string;
}

export interface ProviderAdapter {
  name(): string;
  healthz(): Promise<boolean>;
  invoke(req: InvokeRequest): Promise<InvokeResponse>;
}

File: contracts/adapter_interface/provider_adapter.py
# adapter_interface/provider_adapter.py
from __future__ import annotations
from typing import Protocol, TypedDict, List, Optional, Dict, Any

class InvokeInput(TypedDict, total=False):
    messages: List[Dict[str, Any]]
    system: str
    tools: List[Dict[str, Any]]

class InvokeRequest(TypedDict, total=False):
    provider: str
    operation: str   # 'chat' | 'embed' | 'vision' | 'audio'
    model: str
    budget_tokens: int
    temperature: float
    input: InvokeInput
    metadata: Dict[str, Any]

class Usage(TypedDict, total=False):
    input_tokens: int
    output_tokens: int
    cost: float

class InvokeOutput(TypedDict, total=False):
    messages: List[Dict[str, Any]]
    text: str
    tool_calls: List[Dict[str, Any]]

class InvokeResponse(TypedDict, total=False):
    output: InvokeOutput
    usage: Usage
    provider: str
    model: str

class ProviderAdapter(Protocol):
    def name(self) -> str: ...
    def healthz(self) -> bool: ...
    def invoke(self, req: InvokeRequest) -> InvokeResponse: ...
